### DISCUSSION TITLE
Could not consume arg: --config_path

### DISCUSSION THREAD

**Discussion by imadgohar** [obss/sahi#712] [Q&A] [answered]
Hi, 
I am using detecctron2 model with sahi, installation is okk and i can run the script but in output i just getting visual file with images but no detections at all.

and at the end giving me this error:
Prediction results are successfully exported to runs/predict/exp2
**ERROR: Could not consume arg: --config_path**
Usage: sahi predict --source /test/ --model_type detectron2 --model_path /content/drive/MyDrive/SAHI/Detectron2-SAHI/model_final.pth --config_path /content/drive/MyDrive/SAHI/Detectron2-SAHI/config2.yaml --slice_height 640 --slice_width 640

Is it something wrong with the config file? or something else I am missing?

this is how i am creating the config file from the tutorial given:
# create desired Detectron2 config
from detectron2.config import get_cfg
cfg = get_cfg()
cfg.MODEL.ROI_HEADS.NUM_CLASSES = 5 
cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = 0.5
# export config as yaml
from sahi.utils.detectron2 import export_cfg_as_yaml 
export_cfg_as_yaml(cfg, export_path='config.yaml')

**Comment by fcakyon** [comment, marked as answer]
Hello,

It should be —model_config_path instead of —config_path.

**Reply by imadgohar** [reply]
Thanks @fcakyon for your quick response. I am trying it, I just followed the tutorial and it is like this there:
sahi predict --source image_dir/ --model_type detectron2 --model_path weight.pt --config_path config.yaml --slice_height 512 --slice_width 512

**Reply by fcakyon** [reply]
Can you post the link of tutorial so that I can fix the typo?

**Reply by imadgohar** [reply]
Sure, here it is:
https://medium.com/codable/perform-sliced-inference-and-detailed-error-analysis-using-detectron2-models-37063e33073f

**Reply by imadgohar** [reply]
After executing the command I am only getting Visual file in runs, no ground truth visual and result.json file is there.

**Reply by fcakyon** [reply]
In order to get ground truth visuals and result.json, you have to provide coco dataset json file to the predict command.

It is detailed under `Error Analysis Plots/Metrics` title in the [medium blog](https://medium.com/codable/perform-sliced-inference-and-detailed-error-analysis-using-detectron2-models-37063e33073f) and also detailed under [this discussion](https://github.com/obss/sahi/discussions/622).

**Reply by imadgohar** [reply]
![resu](https://user-images.githubusercontent.com/45845910/199668342-d08a6227-95cb-406f-9a9e-de5f462654a9.PNG)
Why I am getting 0 in all results. Tried the id's from 0,1,2,3 to 1,2,3,4 both way same results.

**Reply by fcakyon** [reply]
category ids should start from 0

what are your inference params?

can you post your results.json and dataset.json?

**Reply by imadgohar** [reply]
predict(
    model_type=model_type,
    model_path=model_path,
    model_device=model_device,
    dataset_json_path = dataset_json_path,
    source=source_image_dir,
    slice_height=slice_height, #640
    slice_width=slice_width, #640
    overlap_height_ratio=overlap_height_ratio, #0.2
    overlap_width_ratio=overlap_width_ratio,  #0.2
    postprocess_match_metric="IOU",
    postprocess_type = 'NMS',
    postprocess_match_threshold = 0.5,
    model_confidence_threshold=0.01
)

Yes, I am sharing it

**Reply by fcakyon** [reply]
have you tried decreasing `cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST` to 0.01 when creating the detectron2 config file?

**Reply by imadgohar** [reply]
Tried same results. Not working
result.json->
[{"image_id":1,"bbox":[606.77099609375,2451.495147705078,0.00189208984375,236.50485229492188],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":0},{"image_id":1,"bbox":[5256.604187011719,2330.0,0.156494140625,640.0],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":100},{"image_id":1,"bbox":[5163.680358886719,2857.2940673828125,0.0078125,112.7059326171875],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":0},{"image_id":1,"bbox":[5085.2724609375,2330.0,1.02783203125,640.0],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":657},{"image_id":1,"bbox":[5017.944427490234,2945.17041015625,0.001190185546875,24.82958984375],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":0},{"image_id":1,"bbox":[3750.1458740234375,2869.0701293945312,1.134796142578125,100.92987060546875],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":114},{"image_id":1,"bbox":[1838.07666015625,2330.0,0.320953369140625,640.0],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":205},{"image_id":1,"bbox":[4917.563446044922,2640.1520385742188,0.0460205078125,47.84796142578125],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":2},{"image_id":1,"bbox":[4720.1004638671875,2048.0,0.40234375,640.0],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":257},{"image_id":1,"bbox":[4694.1192626953125,2048.0,0.17510986328125,640.0],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":112},{"image_id":1,"bbox":[4493.251647949219,2165.2080307006836,0.00244140625,49.00293731689453],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":0},{"image_id":1,"bbox":[3503.03662109375,2129.4319076538086,0.798828125,27.227005004882812],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":21},{"image_id":1,"bbox":[3134.51708984375,2048.0,0.76220703125,640.0],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":487},{"image_id":1,"bbox":[881.08935546875,512.0,0.1240234375,640.0],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":79},{"image_id":1,"bbox":[5213.019348144531,1932.611572265625,0.0048828125,243.388427734375],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":1},{"image_id":1,"bbox":[2571.5861206054688,1536.0,0.33270263671875,640.0],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":212},{"image_id":1,"bbox":[2409.8589477539062,1536.0,0.526702880859375,640.0],"score":1.0,"category_id":1,"category_name":"1","segmentation":[],"iscrowd":0,"area":337},{"image_id":1,"bbox":[2454.6143188476562,1997.817138671875,0.437896728515625,178.182861328125],"score":1.0,"category_id":1,"category_name":"1",

datset.json ->
"annotations":[{"iscrowd":0,"image_id":1,"bbox":[3306,899,39,54],"segmentation":[],"category_id":1,"id":1,"area":2106},{"iscrowd":0,"image_id":1,"bbox":[3089,1350,96,89],"segmentation":[],"category_id":4,"id":2,"area":8544},{"iscrowd":0,"image_id":2,"bbox":[898,217,74,62],"segmentation":[],"category_id":1,"id":3,"area":4588},{"iscrowd":0,"image_id":2,"bbox":[1047,449,59,52],"segmentation":[],"category_id":4,"id":4,"area":3068},{"iscrowd":0,"image_id":2,"bbox":[2569,1118,21,22],"segmentation":[],"category_id":0,"id":5,"area":462},{"iscrowd":0,"image_id":2,"bbox":[2746,1047,20,21],"segmentation":[],"category_id":0,"id":6,"area":420},{"iscrowd":0,"image_id":2,"bbox":[1821,2062,23,27],"segmentation":[],"category_id":0,"id":7,"area":621},{"iscrowd":0,"image_id":3,"bbox":[1185,1184,250,156],"segmentation":[],"category_id":0,"id":8,"area":39000},{"iscrowd":0,"image_id":4,"bbox":[3120,64,85,146],"segmentation":[],"category_id":0,"id":9,"area":12410},{"iscrowd":0,"image_id":5,"bbox":[1364,343,51,61],"segmentation":[],"category_id":1,"id":10,"area":3111},{"iscrowd":0,"image_id":5,"bbox":[1455,569,46,44],"segmentation":[],"category_id":4,"id":11,"area":2024},{"iscrowd":0,"image_id":5,"bbox":[2829,1816,23,25],"segmentation":[],"category_id":0,"id":12,"area":575},{"iscrowd":0,"image_id":5,"bbox":[3074,1858,24,24],"segmentation":[],"category_id":0,"id":13,"area":576},{"iscrowd":0,"image_id":5,"bbox":[2060,2279,22,19],"segmentation":[],"category_id":0,"id":14,"area":418},{"iscrowd":0,"image_id":6,"bbox":[4135,1490,225,185],"segmentation":[],"category_id":1,"id":15,"area":41625},{"iscrowd":0,"image_id":7,"bbox":[2489,766,17,16],"segmentation":[],"category_id":0,"id":16,"area":272},{"iscrowd":0,"image_id":7,"bbox":[2494,806,13,16],"segmentation":[],"category_id":0,"id":17,"area":208},{"iscrowd":0,"image_id":7,"bbox":[2535,1245,21,33],"segmentation":[],"category_id":1,"id":18,"area":693},{"iscrowd":0,"image_id":8,"bbox":[1414,410,317,343],"segmentation":[],"category_id":4,"id":19,"area":108731},{"iscrowd":0,"image_id":8,"bbox":[2047,376,55,57],"segmentation":[],"category_id":1,"id":20,"area":3135},{"iscrowd":0,"image_id":8,"bbox":[2321,704,54,58],"segmentation":[],"category_id":1,"id":21,"area":3132},{"iscrowd":0,"image_id":8,"bbox":[2625,1109,73,67],"segmentation":[],"category_id":1,"id":22,"area":4891},{"iscrowd":0,"image_id":8,"bbox":[2912,1441,74,67],"segmentation":[],"category_id":1,"id":23,"area":4958},{"iscrowd":0,"image_id":8,"bbox":[3211,1815,73,68],"segmentation":[],"category_id":1,"id":24,"area":4964},{"iscrowd":0,"image_id":8,"bbox":[3525,2194,63,59],"segmentation":[],"category_id":1,"id":25,"area":3717},{"iscrowd":0,"image_id":8,"bbox":[3916,2678,79,85],"segmentation":[],"category_id":1,"id":26,"area":6715},{"iscrowd":0,"image_id":9,"bbox":[1592,1410,40,39],"segmentation":[],"category_id":1,"id":27,"area":1560},{"iscrowd":0,"image_id":10,"bbox":[812,1087,48,41],"segmentation":[],"category_id":1,"id":28,"area":1968},{"iscrowd":0,"image_id":11,"bbox":[3239,739,23,22],"segmentation":[],"category_id":0,"id":29,"area":506},{"iscrowd":0,"image_id":11,"bbox":[3454,806,24,23],"segmentation":[],"category_id":0,"id":30,"area":552},{"iscrowd":0,"image_id":11,"bbox":[4500,1133,30,30],"segmentation":[],"category_id":2,"id":31,"area":900},{"iscrowd":0,"image_id":12,"bbox":[2030,574,14,39],"segmentation":[],"category_id":4,"id":32,"area":546},{"iscrowd":0,"image_id":12,"bbox":[2230,1457,40,188],"segmentation":[],"category_id":3,"id":33,"area":7520},{"iscrowd":0,"image_id":12,"bbox":[2296,2559,19,18],"segmentation":[],"category_id":0,"id":34,"area":342},{"iscrowd":0,"image_id":12,"bbox":[2574,2462,28,27],"segmentation":[],"category_id":0,"id":35,"area":756},{"iscrowd":0,"image_id":12,"bbox":[2702,2659,29,27],"segmentation":[],"category_id":0,"id":36,"area":783},{"iscrowd":0,"image_id":13,"bbox":[886,977,42,35],"segmentation":[],"category_id":1,"id":37,"area":1470},{"iscrowd":0,"image_id":14,"bbox":[2814,516,24,26],"segmentation":[],"category_id":0,"id":38,"area":624},{"iscrowd":0,"image_id":14,"bbox":[2789,778,29,37],"segmentation":[],"category_id":0,"id":39,"area":1073},{"iscrowd":0,"image_id":14,"bbox":[2939,1376,33,32],"segmentation":[],"category_id":0,"id":40,"area":1056},{"iscrowd":0,"image_id":14,"bbox":[2712,1572,46,75],"segmentation":[],"category_id":1,"id":41,"area":3450},{"iscrowd":0,"image_id":14,"bbox":[2670,1607,48,159],"segmentation":[],"category_id":3,"id":42,"area":7632},{"iscrowd":0,"image_id":15,"bbox":[3015,1358,94,57],"segmentation":[],"category_id":4,"id":43,"area":5358},{"iscrowd":0,"image_id":15,"bbox":[2779,1070,394,466],"segmentation":[],"category_id":1,"id":44,"area":183604},{"iscrowd":0,"image_id":16,"bbox":[5090,414,90,91],"segmentation":[],"category_id":0,"id":45,"area":8190},{"iscrowd":0,"image_id":17,"bbox":[2281,1660,59,57],"segmentation":[],"category_id":3,"id":46,"area":3363},{"iscrowd":0,"image_id":17,"bbox":[2033,1858,69,64],"segmentation":[],"category_id":3,"id":47,"area":4416},{"iscrowd":0,"image_id":17,"bbox":[1817,2083,70,105],"segmentation":[],"category_id":1,"id":48,"area":7350},{"iscrowd":0,"image_id":17,"bbox":[1738,2098,85,78],"segmentation":[],"category_id":3,"id":49,"area":6630},{"iscrowd":0,"image_id":18,"bbox":[2231,2553,465,417],"segmentation":[],"category_id":4,"id":50,"area":193905},

**Reply by fcakyon** [reply]
its worth to give it a try

**Reply by imadgohar** [reply]
> its worth to give it a try

Not working tried.

**Reply by imadgohar** [reply]
With YOLO results are fine although it is lower than actual yolov5 results but close to that. With detectron2 it's not working.

**Reply by fcakyon** [reply]
how is the detection visuals? which detectron2 model are you using? your detectron2 outputs don't make any sense. are you sure your model is properly trained?

**Reply by imadgohar** [reply]
I am using RetinaNet from detectron2. Training and validation is going well. I just checked to run your colab notebook for Detectron2 + SAHI and tried single image but not detecting anything. Getting this warning when passing config file.

WARNING:sahi.models.detectron2:tuple index out of range

**Reply by fcakyon** [reply]
Can you provide your model and config file so that i can recreate the issue?

**Reply by imadgohar** [reply]
![capt](https://user-images.githubusercontent.com/45845910/199909515-7a08cc3b-3c4d-4113-a842-d91e86e14f08.PNG)

**Reply by imadgohar** [reply]
> config file so

Yes, How I can share it with you?

**Reply by fcakyon** [reply]
Maybe you upload it to google drive and send me a public link? You can either paste the link here or send me a private message from my [linkedin](https://www.linkedin.com/in/fcakyon/) or [twitter](https://twitter.com/fcakyon) accounts.

**Comment by fcakyon** [comment]
Normally there is not issue with sahi+detectron2 support: https://github.com/obss/sahi/blob/main/demo/inference_for_detectron2.ipynb

This issue is most probably caused by your detection2 model.

**Reply by imadgohar** [reply]
On the validation set I got around @mAP50 80, what should be the possible reason I don't know.

**Comment by fcakyon** [comment]
@imadgohar you have to set num_classes as `cfg.MODEL.RETINANET.NUM_CLASSES` for the retinanet model, have you tried that way instead of `cfg.MODEL.ROI_HEADS.NUM_CLASSES` ?

**Reply by imadgohar** [reply]
Yes, I am considering it. Let me train the model again.
